[RFC] ppc64le enablement

CrystalGamma opened the pull request:
This PR enables the ppc64le architecture (and, theoretically, also Big Endian ppc64, but it's broken).

Building the ppc64le-glibc bootstrap files was successfully tested on x86-64 and ppc64le-glibc (building them on ppc64le-musl fails because of some doubly-defined symbols).
Building the ppc64le-musl bootstrap files was tested on x86-64, ppc64le-glibc and ppc64le-musl.
This obviously includes building the ppc64le-glibc and ppc64le-musl stdenv.

I have written a short roadmap/overview for what's in the pipe for NixOS/nixpkgs on POWER: https://crystalgamma.github.io/nixpkgs-prs.html .

As for this particular PR, should the changes be broken up into smaller commits or is this fine?

Commit:
ppc64le enablement

Ericson2314 commented:
See also https://github.com/NixOS/nixpkgs/pull/44092, which I had every intention to merge once the remaining issues were fixed.

Commit:
gcc, glibc: properly handle __float128

CrystalGamma commented:
I didn't know a previous PR for POWER existed …

ppc64 is just one bit flip in the MSR (machine status register) away from ppc64le (just like 32-bit mode), so every ppc64le computer could (and I have every intention to at least try) run ppc64 code (even in the same OS if the kernel supports it, which Linux doesn't AFAIK). (The author of the previous PR was right in that the firmware of older ppc64 hardware and pretty much all ppc32 hardware could only boot big-endian OSs) In fact OpenPower systems which are generally marketed as ppc64le hardware actually boot in big-endian mode AFAIK, and then switch to whatever the OS is compiled for. That's why I introduced a more general isPower inspection predicate (and endian-specific options can then check isBigEndian or isLittleEndian) for platforms.

I did see a GCC config option (--with-long-double-128) in that PR which could have saved me about 50h of my life had I known about it earlier. Using that allowed me to remove some hackery in glibc. (I can squash before merging, if that's desired)

You may note that this PR also allows you to build a musl stdenv, which the earlier one didn't, AFAICS.

Ericson2314 commented on pkgs/development/compilers/gcc/7/default.nix:
```diff
@@ -260,7 +261,12 @@ stdenv.mkDerivation ({
+    # the float128 detection in cross-gcc and glibc is a little busted. since glibc doesn't compile without float128 support, fix it.
+    sed 's/-mfloat128/-mfloat128-type -mfloat128/g' -i libgcc/configure
+    # this file also includes totally unneeded headers which are not available in cross-compilation. remove them.
+    sed -E 's/#include <(string|stdlib|ctype).h>//' -i libgcc/config/rs6000/float128-ifunc.c
```
Probably good to put the comment outside the string, so changing the explanation doesn't cause rebuilds.

CrystalGamma replied:
Makes sense. So like this?

```
# the float128 detection in cross-gcc and glibc is a little busted. since glibc doesn't compile without float128 support, fix it.
+ "\nsed 's/-mfloat128/-mfloat128-type -mfloat128/g' -i libgcc/configure"
# this file also includes totally unneeded headers which are not available in cross-compilation. remove them.
+"\nsed -E 's/#include <(string|stdlib|ctype).h>//' -i libgcc/config/rs6000/float128-ifunc.c"
```

Anyway, it turns out this hack was also obsoleted by just using `--with-long-double-128 --enable-softfloat` and I just overlooked it when cleaning up …

Ericson2314 commented on pkgs/development/compilers/gcc/7/default.nix:
```diff
@@ -325,6 +331,10 @@ stdenv.mkDerivation ({
     ++ optional (targetPlatform.isAarch64) "--enable-fix-cortex-a53-843419"
+    ++ optionals (targetPlatform.isPower && targetPlatform.libc == "glibc") [
+      "--with-long-double-128"
+      "--enable-softfloat"
```
It would be nice to leverage https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/gcc/common/platform-flags.nix for this

CrystalGamma replied:
AFAICS platform-flags only has `--with-xyz=abc` with a select few 'xyz'. Unless you'd like me to extend it …

(Also, __float128 being required is only for the combination glibc, little endian, 64 bits … I might actually want to check endianness and word size too …)

Ericson2314 commented on pkgs/development/compilers/gcc/8/default.nix:
```diff
@@ -316,6 +317,10 @@ stdenv.mkDerivation ({
+    ++ optionals (targetPlatform.isPower && targetPlatform.libc == "glibc") [
+      "--with-long-double-128"
+      "--enable-softfloat"
+    ]
```
Don't forget to update pkgs/development/compilers/gcc/snapshot/default.nix too

Ericson2314 replied:
Yeah it's some arbitrary vcs revision. Should be newer than the rest so it would become the next release version, GCC 9 here.

Ericson2314 commented on pkgs/stdenv/default.nix:
```diff
@@ -47,6 +47,7 @@ in
     "aarch64-linux" = stagesLinux;
     "mipsel-linux" = stagesLinux;
     "powerpc-linux" = /* stagesLinux */ stagesNative;
+    powerpc64le-linux = stagesLinux;
```
Use `"` here and other such places for consistency

Ericson2314 replied:
I think you got them all but the one I commented on, haha :)

Commit:
lib/systems, stdenv: syntax cleanup

Commit:
gcc7: remove ugly hack

Commit:
gcc: add/update __float128 flags

CrystalGamma commented:
So, I made the quotes consistent, removed an obsolete hack and updated the flags for gcc7, gcc8 and gcc-snapshot. I am not certain how to best abstract the flags into the platform definition using gcc/common/platform-flags.nix though.

Ericson2314 commented:
So for the 128-bit float, I'd add 
```
optional (targetPlatform.isPower && targetPlatform.libc == "glibc" && targetPlatform.is64bit && target.isLittleEndian) "--with-long-double-128"
```
the the concatentated list in https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/gcc/common/platform-flags.nix#L12

For `--enable-softfloat`, I'd need to know what that is used for. That is a multilib flag, and our wrapper scripts don't really play well with multilib. [I am working on packaging the GCC runtime libraries separately for a "nix-driven multilib", side-stepping this restriction.] So I think it be better to use `--with-float` and friends to force float support one way or the other, if possible.

Commit:
stdenv: add another pair of quotes for consistency

Commit:
gcc: move __float128 flag for ppc64le-glibc into common/platform-flags.nix

CrystalGamma commented:
I blindly copied the `--enable-softfloat` flag from the earlier PR. I assumed it was necessary because I had seen linker errors referencing 128-bit float emulation routines when dabbling in cross-building glibc. Thanks for making me challenge that hypothesis :D

I was thinking that maybe passing the `--with-long-double-128` flag to all versions of GCC may cause failures in older versions. But since versions pre-7 didn't support little-endian ppc64, I guess it wouldn't matter either way …

Ericson2314 commented:
Sounds good! Did `--enable-softfloat` work for cross then? Is this ready to merge?

CrystalGamma commented:
It did. Should I squash these changes before we merge?

Ericson2314 commented:
Glad to hear it! I can just "squash and merge", but if you want to divide things into more and 1 commit that's fine too. I'll check this again when I wake up.

CrystalGamma commented:
No, squashing is fine by me. Thanks for helping me bring support for more niche architectures to NixOS :)

Ericson2314 commented:
Of course!

dtzWill commented on pkgs/os-specific/linux/musl/default.nix:
```diff
@@ -61,11 +61,12 @@ stdenv.mkDerivation rec {
     configureFlagsArray+=("--syslibdir=$out/lib")
   '';
 
+  CFLAGS="-fstack-protector-strong" + lib.optionalString hostPlatform.isPower " -mlong-double-64";
```
This isn't the same as only setting CFLAGS during configure-- have you checked the output is identical on x86_64? Hopefully doesn't matter but better safe than sorry....

dtzWill replied:
And just curious about why this flag is needed, but I haven't read through the discussion/your website yet so apologies if all this has been addressed.

CrystalGamma replied:
The rationale here was that since CFLAGS might contain a space now, it might show up as two options, one of which ./configure will not understand

CrystalGamma replied:
The original problem is that long double defaults to 128-bits on ppc64le (even without `--with-long-double-128` which is required for glibc interestingly) but musl wants 64-bit long double.

dtzWill commented on pkgs/stdenv/linux/bootstrap-tools/scripts/unpack-bootstrap-tools.sh:
```diff
@@ -9,6 +9,9 @@ echo Patching the bootstrap tools...
    LD_BINARY=$out/lib/ld.so.?
+elif test -f $out/lib/ld64.so.?; then
+   # ppc64(le)
+   LD_BINARY=$out/lib/ld64.so.?
```
Does the musl unpack script need similar updating? Not sure what it's called in that situation.

CrystalGamma replied:
It certainly works without any additional changes, so I guess not …

dtzWill reviewed (commented):
Excited to see this landing, here's some feedback.  Will read through the discussion/your website post more thoroughly in a bit.
